### PR TITLE
Pull #13176: Fail linkcheck on internal broken site link

### PR DESCRIPTION
**Pull request by stoyanK7**
Extends script to fail on broken links between xdocs.

---
## EDIT: I had excluded any checks in and to `apidocs/` when I ran those scripts. Therefore some 1300 errors that were present before do not show.

I broke two links before running the experiments below.
```diff
modified: src/xdocs/config_filters.xml
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@ src/xdocs/config_filters.xml:122 @
          (<code>&lt;module name=&quot;TreeWalker&quot;/&gt;</code>) and only
          applies to checks which are also defined within this module.
          To filter non-TreeWalker checks like <code>RegexpSingleline</code>,
-          a <a href="config_filters.html#SuppressWithPlainTextCommentFilter">
+          a <a href="asdasdfilters.html#SuppressWithPlainTextCommentFilter">
          SuppressWithPlainTextCommentFilter</a> or similar filter must be used.
        </p>
      </subsection>
      <subsection name="Notes" id="SuppressionCommentFilter_Notes">
        <p>
          <code>offCommentFormat</code> and <code>onCommentFormat</code> must have equal
-          <a href="https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/regex/Matcher.html#groupCount()">
+          <a href="https://docs.e.com/en/java/javase/11/docs/api/java.base/java/util/regex/Matcher.html#groupCount()">
            paren counts</a>.
        </p>
        <p>

```

### Before changes
The script greps `externalLink`, finds broken external link to oracle but not the internal one.
```sh
$ ./.ci/run-link-check-plugin.sh
...
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:52 min
[INFO] Finished at: 2023-06-07T10:37:42+02:00
[INFO] ------------------------------------------------------------------------
------------ grep of linkcheck.html--BEGIN
<td><i><a class="externalLink" href="https://docs.e.com/en/java/javase/11/docs/api/java.base/java/util/regex/Matcher.html#groupCount()">https://docs.e.com/en/java/javase/11/docs/api/java.base/java/util/regex/Matcher.html#groupCount()</a>: java.net.UnknownHostException : docs.e.com</i></td></tr></table></td></tr>
------------ grep of linkcheck.html--END
Exit code:1
```

In the report the 2 errors are shown.
image

### After changes

```sh
$ ./.ci/run-link-check-plugin.sh
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:50 min
[INFO] Finished at: 2023-06-07T10:52:41+02:00
[INFO] ------------------------------------------------------------------------
------------ grep of linkcheck.html--BEGIN
<td><i><a href="asdasdfilters.html#SuppressWithPlainTextCommentFilter">asdasdfilters.html#SuppressWithPlainTextCommentFilter</a>: doesn't exist.</i></td></tr>
<td><i><a class="externalLink" href="https://docs.e.com/en/java/javase/11/docs/api/java.base/java/util/regex/Matcher.html#groupCount()">https://docs.e.com/en/java/javase/11/docs/api/java.base/java/util/regex/Matcher.html#groupCount()</a>: java.net.UnknownHostException : docs.e.com</i></td></tr></table></td></tr>
------------ grep of linkcheck.html--END
Exit code:2

```

### CODE REVIEW

**Comment by stoyanK7** [comment]
@rnveach @nrmancuso @romani We need this to make the linkcheck job work as we want - catch broken links between site pages.

**Comment by stoyanK7** [comment]
@nrmancuso https://app.circleci.com/pipelines/github/checkstyle/checkstyle/18778/workflows/49905029-0d70-42cf-b1c6-4c69d50f6a78/jobs/302519?invite=true#step-103-1683

Uhh, that's different. Shall I just make an exclusion list for those and open an issue to deal with later? The `apidocs/` failures are already mentioned at https://github.com/checkstyle/checkstyle/issues/11572.

**Comment by nrmancuso** [comment]
@stoyanK7 i am not sure that I understand what’s going on here. Was CI green until we added one bad link, then we got 1306 unexpected failures?

**Comment by stoyanK7** [comment]
@nrmancuso Those failures were always present in the report if you check it at `target/site/linkcheck.html`. The script greps only `externalLink`s and didn't see them until now.

**Comment by nrmancuso** [comment]
Then I don’t understand

> Uhh, that's different

**Comment by stoyanK7** [comment]
@nrmancuso Just checked. I didn't see these errors from the script locally because I had excluded checks in `apidocs/` in `pom.xml`. That's my bad.

**Comment by stoyanK7** [comment]
We have to suppress those for now.

**Comment by nrmancuso** [comment]
> We have to suppress those for now.

Please do, make sure we have `until ...11572` comment on suppressions

File: .ci/run-link-check-plugin.sh [hunk 1]
```diff
@@ -9,12 +9,25 @@ mvn --version
...
+# Suppressions are already sorted
+sort .ci-temp/linkcheck-errors.txt > .ci-temp/linkcheck-errors-sorted.txt
+
+# Suppressions exist until https://github.com/checkstyle/checkstyle/issues/11572
+diff config/linkcheck-suppressions.txt .ci-temp/linkcheck-errors-sorted.txt || true
 echo "------------ grep of linkcheck.html--END"
-RESULT=$(grep externalLink target/site/linkcheck.html | grep -v 'Read timed out' | wc -l)
+
+RESULT=$(diff -y --suppress-common-lines config/linkcheck-suppressions.txt \
+  .ci-temp/linkcheck-errors-sorted.txt | wc -l)
```
Comment (stoyanK7) [inline .ci/run-link-check-plugin.sh:28]: Taken from https://stackoverflow.com/a/48287203/9553927
> `diff` can do all the first part of the job but no counting; `wc` -l does the rest:

> If someone was wondering: `-y` is a shorthand for `--side-by-side`, which outputs the files in two columns.

**Comment by stoyanK7** [comment]
I believe this is the failure of semaphore(I still have no access to logs):
https://checkstyle.semaphoreci.com/jobs/bee16411-e62e-4873-b5db-2753f6e36eda
```
Content of target/eclipse/report.txt:
Error: LinkageError occurred while loading main class org.eclipse.jdt.internal.compiler.batch.Main
        java.lang.UnsupportedClassVersionError: org/eclipse/jdt/internal/compiler/batch/Main has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0
[ERROR] Command execution failed.
org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)

```

**Comment by nrmancuso** [comment]
>I believe this is the failure of semaphore(I still have no access to logs):

I have PR opened at https://github.com/checkstyle/checkstyle/pull/13201 for this

**Comment by stoyanK7** [comment]
> Please use PR number instead of minor, it is not minor update.

Done.

**Comment by nrmancuso** [comment]
@stoyanK7 please make CI happy, linkcheck is failing

File: .ci/run-link-check-plugin.sh [hunk 2]
```diff
@@ -9,12 +9,26 @@ mvn --version
...
+mkdir -p .ci-temp
+
 echo "------------ grep of linkcheck.html--BEGIN"
 # "grep ... | cat" is required command is running in "set -e" mode and
 # grep could return exit code 1 if nothing is matching
-grep externalLink target/site/linkcheck.html | cat
+grep -E "doesn't exist|externalLink" target/site/linkcheck.html | grep -v 'Read timed out' \
+  |  cat > .ci-temp/linkcheck-errors.txt
+
+sort config/linkcheck-suppressions.txt > .ci-temp/linkcheck-suppressions-sorted.txt
```
Comment (nrmancuso) [inline .ci/run-link-check-plugin.sh:20]: Why not pipe to sort in original grep?

File: .ci/run-link-check-plugin.sh [hunk 3]
```diff
@@ -9,12 +9,26 @@ mvn --version
...
+sort .ci-temp/linkcheck-errors.txt > .ci-temp/linkcheck-errors-sorted.txt
+
+# Suppressions exist until https://github.com/checkstyle/checkstyle/issues/11572
+diff .ci-temp/linkcheck-suppressions-sorted.txt .ci-temp/linkcheck-errors-sorted.txt || true
 echo "------------ grep of linkcheck.html--END"
-RESULT=$(grep externalLink target/site/linkcheck.html | grep -v 'Read timed out' | wc -l)
+
+RESULT=$(diff -y --suppress-common-lines .ci-temp/linkcheck-suppressions-sorted.txt \
+  .ci-temp/linkcheck-errors-sorted.txt | wc -l)
+rm .ci-temp/linkcheck-errors.txt
```
Comment (nrmancuso) [inline .ci/run-link-check-plugin.sh:29]: Why? We have a fresh container every time we execute this, right?

File: .ci/run-link-check-plugin.sh [hunk 4]
```diff
@@ -9,12 +9,26 @@ mvn --version
...
 # grep could return exit code 1 if nothing is matching
-grep externalLink target/site/linkcheck.html | cat
+grep -E "doesn't exist|externalLink" target/site/linkcheck.html | grep -v 'Read timed out' \
+  |  cat > .ci-temp/linkcheck-errors.txt
+
+sort config/linkcheck-suppressions.txt > .ci-temp/linkcheck-suppressions-sorted.txt
+sort .ci-temp/linkcheck-errors.txt > .ci-temp/linkcheck-errors-sorted.txt
+
+# Suppressions exist until https://github.com/checkstyle/checkstyle/issues/11572
+diff .ci-temp/linkcheck-suppressions-sorted.txt .ci-temp/linkcheck-errors-sorted.txt || true
```
Comment (nrmancuso) [inline .ci/run-link-check-plugin.sh:24]: Why `|| true`?

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Items:

File: .ci/run-link-check-plugin.sh (hunk 4 above)
Comment (nrmancuso) [reply .ci/run-link-check-plugin.sh:24]: Disregard, comment confused me a bit.

File: .ci/run-link-check-plugin.sh (hunk 2 above)
Comment (stoyanK7) [reply .ci/run-link-check-plugin.sh:20]: The idea was readability.

Comment (stoyanK7) [reply .ci/run-link-check-plugin.sh:20]: Looks fine with pipe in original grep too though. Done.

File: .ci/run-link-check-plugin.sh (hunk 3 above)
Comment (stoyanK7) [reply .ci/run-link-check-plugin.sh]: Uhh, I got curious if other scripts remove .ci-temp files, found some and decided to do it for some reason. Disregard. Removed.

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Last item from me:

File: .ci/run-link-check-plugin.sh [hunk 5]
```diff
@@ -9,12 +9,24 @@ mvn --version
...
 mvn -e --no-transfer-progress clean site -Dcheckstyle.ant.skip=true -DskipTests -DskipITs \
    -Dpmd.skip=true -Dspotbugs.skip=true -Djacoco.skip=true -Dcheckstyle.skip=true
+mkdir -p .ci-temp
+
 echo "------------ grep of linkcheck.html--BEGIN"
 # "grep ... | cat" is required command is running in "set -e" mode and
 # grep could return exit code 1 if nothing is matching
-grep externalLink target/site/linkcheck.html | cat
+grep -E "doesn't exist|externalLink" target/site/linkcheck.html | grep -v 'Read timed out' \
+  | cat | sort > .ci-temp/linkcheck-errors-sorted.txt
```
Comment (nrmancuso) [inline .ci/run-link-check-plugin.sh:18]: Do we need `cat` now?

**Comment by stoyanK7** [comment]
I fixed a missed broken link from split https://github.com/checkstyle/checkstyle/pull/13103

**Comment by romani** [comment]
Please fix CI
https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/47324021/job/5b0l4bggb4a59qef#L113

`[ERROR] [checkstyle] [ERROR] C:\projects\checkstyle\config\linkcheck-suppressions.txt:1: File does not end with a newline. [NewlineAtEndOfFile]`

**Review by nrmancuso — APPROVED** [review]
I am good as CI passes.

**Comment by stoyanK7** [comment]
@romani This one is good to go.

**Review by romani — APPROVED** [review]
Ok to merge

**Pull request merged**